File: 240428/퀵 정렬 구현/implement-quick-sort.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    public static void main(String[] args) {
        int n = sc.nextInt();
        int[] arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = sc.nextInt();
        }

        quickSort(arr, 0, n-1);

        for(int item: arr){
            System.out.print(item + " ");
        }
    }

    static void quickSort(int[] arr, int low, int high) {
        if(low < high){
            int pos = partition(arr, low, high);

            quickSort(arr, low, pos -1);
            quickSort(arr, pos+1, high);
        }
    }

    static int partition(int[] arr, int low, int high) {
        int p = arr[high];

        // 먼저 오른쪽으로 가는 인덱스
        int r = low;
        // 왼쪽에서 따라가는 인덱스
        int b = low-1;

        for(int i = low; i < high; i++) {
            if(arr[i] < p) {
                b++;
                swap(arr, b, i);
            }
        }
        b++;
        swap(arr, b, high);

        return b;
    }

    static void swap(int[] arr, int x, int y) {
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }
}

File: 240428/퀵 정렬 구현/implement-quick-sort.kt
import java.util.*

fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val arr = IntArray(n) {sc.nextInt()}

    // quick sort
    quickSort(arr, 0, n-1)


    for(item in arr) {
        print("$item ")
    }

}

// quickSort
fun quickSort(arr: IntArray, low: Int, high: Int){
    if(low < high){
        val pos = partition(arr, low, high)

        quickSort(arr, low, pos-1)
        quickSort(arr, pos+1, high)
    }
}

fun partition(arr: IntArray, low:Int, high: Int): Int{
    val p = arr[high]

    var r = low // 먼저 오른쪽으로 확인하는 인덱스
    var b = low -1 // 왼쪽에서 따라가는 인덱스

    for(r in low .. high-1){
        if(arr[r] < p){
            b++
            swap(arr, r, b)
        }
    }
    b++
    swap(arr, b, high)
    return b
}

fun swap(arr: IntArray, r: Int, b: Int){
    val temp = arr[r]
    arr[r] = arr[b]
    arr[b] = temp
}